### DISCUSSION TITLE
Welcome to ZoneTree.FullTextSearch Discussions!

### DISCUSSION THREAD

**Discussion by koculu** [ZoneTree/ZoneTree.FullTextSearch#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by sgf** [comment]
Inverted index seems to be a key technology in search engines.
The Bitmap algorithm seems to be an important data structure. RoaringBitmap seems to be a data structure that is widely used in this domain (Probably the most advanced bitmap algorithm.).

If possible, you can use RoaringBitmap. Regarding the application of this algorithm, the existing search engines are:
https://github.com/mgholam/hOOt

However, the .net ecosystem currently lacks high-quality implementations of this algorithm. And I don't have the resources (time) to implement this algorithm in my spare time, mainly because there are too many things to do and I am simply too busy (Mainly due to livelihood (work)).

I've proposed this in an open source repository [TheAlgorithms/C-Sharp](https://github.com/TheAlgorithms/C-Sharp/issues/461) ,but I don't think they have any plans to implement it short time soon.

**Comment by koculu** [comment]
Thanks for sharing this.
According to the [project docs](https://www.codeproject.com/Articles/224722/hOOt-full-text-search-engine), indexing is quite slower with hOOt than ZoneTree.FullTextSearch.  

| **Search Engine**      | **Words Inserted** | **Time Taken** | **Words Per Second**      | **Speed Difference**      |
|------------------------|--------------------|----------------|---------------------------|---------------------------|
| **hOOt**               | 290,000            | 56.7 seconds   | ≈ 5,114 words/second       | Baseline                   |
| **ZoneTree.FTS**       | 27,000,000         | 54.8 seconds   | ≈ 492,701 words/second     | ≈ 96.3 times faster        |

ZoneTree inserts approximately 93 times more words in almost the same time, making it 96.3 times faster or 9530% more efficient.

**Comment by sgf** [comment]
I mentioned hOOt not because of how good its implementation is, but mainly to introduce RoaringBitmap.

If it is a search engine, the current mainstream search engines are ES(java) based on lucene. meilisearch(rust), redisearch(c on redis), etc.

**Comment by sgf** [comment]
pure HHD, 67612 records,my creating index takes about 50 mins.


How concurrency is supported by this search engine.

Whether fuzzy queries are supported.

Whether custom tokenizers are supported.

Is it possible to support distributed in the future.

**Reply by koculu** [reply]
What is the total word count?
Could you share the stats using the playground app?
Please also share your CPU, Memory and available physical memory information.

ZoneTree itself provides API to replicate or distribute the data.
ZoneTree and ZoneTree.FullTextSearch are libraries and they can be consumed by a server application that can distribute or replicate the data.

Your other questions are already answered before. 
- Fuzzy support is not planned right now.
- Custom Tokenizer is supported and it is explained in the readme doc.
